drivers/adc/ad7768-1/ad77681.c

```diff
@@ -1169,6 +1169,139 @@ int32_t ad77681_apply_gain(struct ad77681_dev *dev,
 	return ret;
 }
 
+/**
+ * Upload sequence for Programmamble FIR filter
+ * @param dev - The device structure.
+ * @param coeffs - Pointer to the desired filter coefficients array to be written
+ * @param num_coeffs - Count of active filter coeffs
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_programmable_filter(struct ad77681_dev *dev,
+				    const float *coeffs,
+				    uint8_t num_coeffs)
+{
+	uint8_t coeffs_buf[4], coeffs_index, check_back = 0, i, address;
+	uint16_t twait;
+	int32_t twos_complement, ret;
+	const uint8_t coeff_reg_length = 56;
+
+	/* Specific keys in the upload sequence */
+	const uint8_t key1 = 0xAC, key2 = 0x45, key3 = 0x55;
+	/* Scaling factor for all coefficients 2^22 */
+	const float coeff_scale_factor = (1 << 22);
+	/* Wait time in uS necessary to access the COEFF_CONTROL and */
+	/* COEFF_DATA registers. Twait = 512/MCLK */
+	twait = (uint16_t)(((512.0) / ((float)(dev->mclk))) * 1000.0) + 1;
+
+
+	/* Set Filter to FIR */
+	ret = ad77681_spi_write_mask(dev,
+				     AD77681_REG_DIGITAL_FILTER,
+				     AD77681_DIGI_FILTER_FILTER_MSK,
+				     AD77681_DIGI_FILTER_FILTER(AD77681_FIR));
+
+	/* Write the first access key to the ACCESS_KEY register */
+	ret |= ad77681_spi_write_mask(dev,
+				      AD77681_REG_ACCESS_KEY,
+				      AD77681_ACCESS_KEY_MSK,
+				      AD77681_ACCESS_KEY(key1));
+
+	/* Write the second access key to the ACCESS_KEY register */
+	ret |= ad77681_spi_write_mask(dev,
+				      AD77681_REG_ACCESS_KEY,
+				      AD77681_ACCESS_KEY_MSK,
+				      AD77681_ACCESS_KEY(key2));
+
+	/* Read the the ACCESS_KEY register bit 0, the key bit */
+	ret |= ad77681_spi_read_mask(dev,
+				     AD77681_REG_ACCESS_KEY,
+				     AD77681_ACCESS_KEY_CHECK_MSK,
+				     &check_back);
+
+	/* Checks key bit, return FAILURE incase not equal to 1 */
+	if ( check_back != 1 )
+		return FAILURE;
+
+	/* Set the initial adress to 0 and enable the  write and coefficient access bits */
+	address = AD77681_COEF_CONTROL_COEFFACCESSEN_MSK
+		  | AD77681_COEF_CONTROL_COEFFWRITEEN_MSK;
+
+
+	/* The COEFF_DATA register has to be filled with 56 coeffs.*/
+	/* In case the number of active filter coefficient is less */
+	/* than 56, zeros will be padded before the desired coeff. */
+	for (i = 0; i < coeff_reg_length; i++) {
+		/* Set the coeff address */
+		ret |= ad77681_spi_reg_write(dev,
+					     AD77681_REG_COEFF_CONTROL,
+					     address);
+
+		/* Wait for Twait uSeconds*/
+		_delay_us(twait);
```

Comment:
In noos we use `udelay()` :
https://github.com/analogdevicesinc/no-OS/blob/115ced3d31d78a705c242b1bab1a3420251047cc/include/delay.h#L54
You should include "delay.h" to use it and add the to the build delay.c from drivers/platform/

Reply:
Done fixing the udelay.